internal/grpc/commander/commander.go

```diff
@@ -48,36 +60,102 @@ func (c *Commander) CommandChannel(server proto.Commander_CommandChannelServer)
 		return err
 	}
 
-	c.logger.Info("New agent connection", "id", id)
-
 	defer func() {
-		c.logger.Info("Removing agent from manager")
-		c.agentMgr.RemoveAgent(id)
+		c.removeConnection(id)
 	}()
 
-	agentLogger := c.logger.WithValues("id", id)
+	idLogger := c.logger.WithValues("id", id)
 
-	agentConn := newConnection(
+	conn := newConnection(
 		id,
-		agentLogger.WithName("connection"),
-		NewBidirectionalChannel(server, agentLogger.WithName("channel")),
+		idLogger.WithName("connection"),
+		NewBidirectionalChannel(server, idLogger.WithName("channel")),
+		c.observedConfig,
 	)
 
-	c.logger.Info("Adding agent to manager")
-	c.agentMgr.AddAgent(agentConn)
+	c.addConnection(conn)
 
-	return agentConn.run(server.Context())
+	return conn.run(server.Context())
 }
 
-// Download will be implemented in a future PR.
-func (c *Commander) Download(_ *proto.DownloadRequest, _ proto.Commander_DownloadServer) error {
-	return nil
+// Download implements the Download method of the Commander gRPC service. An agent invokes this method to download the
+// latest version of the NGINX configuration.
+func (c *Commander) Download(request *proto.DownloadRequest, server proto.Commander_DownloadServer) error {
+	c.logger.Info("Download requested", "message ID", request.GetMeta().GetMessageId())
+
+	id, err := getUUIDFromContext(server.Context())
+	if err != nil {
+		c.logger.Error(err, "failed download")
+		return err
+	}
+
+	conn := c.getConnection(id)
+	if conn == nil {
+		err := fmt.Errorf("connection with id: %s not found", id)
+		c.logger.Error(err, "failed download")
+		return err
+	}
+
+	// TODO: can there be a race condition here?
```

Comment:
@pleshakov I think the connection could update its State to `StateRegistered` as this check is happening. This means we may skip the download here when we shouldn't. I wonder if we should leave this check to the `sendConfig` method. Thoughts?

Reply:
Yes, the agent shouldn't download or upload before registering, but I still don't think we need this check in the commander. We can just return an error from the connection methods.